packages/bsky/src/notifications.ts

```diff
@@ -0,0 +1,239 @@
+  tokens: string[]
+  platform: 1 | 2 // 1 = ios, 2 = android
+  title: string
+  message: string
+  topic: string
+  data?: {
+    [key: string]: string
+  }
+}
+type InsertableNotif = Insertable<Notification>
+
+export class NotificationServer {
+  constructor(public db: Database, public pushEndpoint?: string) {}
+
+  async getUserTokens(did: string) {
+    const userTokens = await this.db.db
+      .selectFrom('notification_push_token')
+      .where('did', '=', did)
+      .selectAll()
+      .execute()
+
+    return userTokens
+  }
+
+  async prepareNotifsToSend(notifications: InsertableNotif[]) {
+    const notifsToSend: PushNotification[] = []
+
+    for (const notif of notifications) {
+      const { did: userDid } = notif
+      const userTokens = await this.getUserTokens(userDid)
+      const attr = await this.getNotificationDisplayAttributes(notif)
+      // if user has no tokens or the post attr cannot be found, skip
+      if (!userTokens || userTokens.length === 0) {
+        continue
+      }
+      if (!attr) {
+        logger.warn(
+          {
+            userDid,
+            notif,
+          },
+          'No notification display attributes found for this notification. Either profile or post data for this notification is missing.',
+        )
+        continue
+      }
+      const { title, body } = attr
+
+      for (const t of userTokens) {
+        const { appId, platform, token } = t
+        if (platform === 'ios' || platform === 'android') {
+          notifsToSend.push({
+            tokens: [token],
+            platform: platform === 'ios' ? 1 : 2,
+            title: title,
+            message: body,
+            topic: appId,
+            data: {
+              reason: notif.reason,
+              recordUri: notif.recordUri,
+              recordCid: notif.recordCid,
+            },
+          })
+        } else {
+          // @TODO: Handle web notifs
+        }
+      }
+    }
+    return notifsToSend
+  }
+
+  /**  1. Get the user's token (APNS or FCM for iOS and Android respectively) from the database
+    User token will be in the format:
+        did || token || platform (1 = iOS, 2 = Android, 3 = Web)
+    2. Send notification to `gorush` server with token
+    Notification will be in the format:
+    "notifications": [
+      {
+        "tokens": string[],
+        "platform": 1 | 2,
+        "message": string,
+        "title": string,
+        "priority": "normal" | "high",
+        "image": string, (Android only)
+        "expiration": number, (iOS only)
+        "badge": number, (iOS only)
+      }
+    ]
+    3. `gorush` will send notification to APNS or FCM
+    4.  store response from `gorush` which contains the ID of the notification
+    5. If notification needs to be updated or deleted, find the ID of the notification from the database and send a new notification to `gorush` with the ID (repeat step 2)
+  */
+  async sendPushNotifications(notifications: PushNotification[]) {
```

Comment:
are we gonna do any batching or rate limiting? worried this is gonna be super noisy to send a push notification on every action

Reply:
Added some in-memory rate-limiting and batching. However, I think I should utilize redis for that. Let me know if that sounds good.

Reply:
Added Redis too/